src/qt/graphics.cpp

```diff
@@ -0,0 +1,1269 @@
+    }
+
+    // sets the transform of this context
+    virtual void SetTransform(const wxGraphicsMatrix& matrix) wxOVERRIDE
+    {
+        const wxQtMatrixData* qmatrix = static_cast<const wxQtMatrixData*>(matrix.GetRefData());
+        m_qtPainter->setTransform(qmatrix->GetQTransform());
+    }
+
+    // gets the matrix of this context
+    virtual wxGraphicsMatrix GetTransform() const wxOVERRIDE
+    {
+        const QTransform& transform = m_qtPainter->transform();
+        wxGraphicsMatrix m;
+        m.SetRefData(new wxQtMatrixData(GetRenderer(), transform));
+        return m;
+    }
+
+    virtual void DrawBitmap(const wxGraphicsBitmap &bmp, wxDouble x, wxDouble y, wxDouble w, wxDouble h) wxOVERRIDE
+    {
+        const QPixmap* pixmap = wxQtBitmapData::GetPixmapFromBitmap(bmp);
+        m_qtPainter->drawPixmap(x, y, w, h, *pixmap);
+    }
+
+    virtual void DrawBitmap(const wxBitmap &bmp, wxDouble x, wxDouble y, wxDouble w, wxDouble h) wxOVERRIDE
+    {
+        DoDrawBitmap(bmp, x, y, w, h, true);
+    }
+
+    virtual void DrawIcon(const wxIcon &icon, wxDouble x, wxDouble y, wxDouble w, wxDouble h) wxOVERRIDE
+    {
+        DoDrawBitmap(icon, x, y, w, h, true);
+    }
+
+    virtual void PushState() wxOVERRIDE
+    {
+        m_qtPainter->save();
+    }
+
+    virtual void PopState() wxOVERRIDE
+    {
+        m_qtPainter->restore();
+    }
+
+    virtual void Flush() wxOVERRIDE
+    {
+    }
+
+    virtual void GetTextExtent(const wxString &str,
+                               wxDouble *width,
+                               wxDouble *height,
+                               wxDouble *descent,
+                               wxDouble *externalLeading) const wxOVERRIDE
+    {
+        wxCHECK_RET(!m_font.IsNull(), wxT("wxQtContext::GetTextExtent - no valid font set"));
+
+        const wxQtFontData* font_data = static_cast<wxQtFontData*>(m_font.GetRefData());
+        m_qtPainter->setFont(font_data->GetFont());
+
+        const QFontMetrics metrics = m_qtPainter->fontMetrics();
+        const QRect bounding_rect = metrics.boundingRect(QString(str));
+
+        if ( width ) *width = bounding_rect.width();
+        if ( height ) *height = bounding_rect.height();
+        if ( descent ) *descent = metrics.descent();
+        if ( externalLeading ) *externalLeading = metrics.leading() - (metrics.ascent() + metrics.descent());
+    }
+
+    virtual void GetPartialTextExtents(const wxString& text, wxArrayDouble& widths) const wxOVERRIDE
+    {
+        wxCHECK_RET(!m_font.IsNull(), wxT("wxQtContext::GetPartialTextExtents - no valid font set"));
+
+        const wxQtFontData* font_data = static_cast<wxQtFontData*>(m_font.GetRefData());
+        m_qtPainter->setFont(font_data->GetFont());
+
+        const QFontMetrics metrics = m_qtPainter->fontMetrics();
+
+        const size_t text_length = text.length();
+
+        widths.Empty();
+        widths.Add(0, text_length);
+
+        for ( size_t i = 1; i < text_length; ++i )
```

Comment:
Is it correct to start iterating at `1` instead of  `0`? 

The ` widths.Add(0, text_length);` above may look as if it for some odd reason puts the stringth length as the first element but in fact it fills the array with `text_length` of `0`s.

Reply:
Have I misinterpreted? My assumption is that the index is how many characters to take the width of (per the docs: from the beginning of text _to_ the corresponding character of text.) So the first index width would be 0 (since you're not including any characters). That might be wrong. It could be argued that having the initial value be 0 is useless, but then having the total width as the last could be argued as useless as well. So I can't reason my way to an answer. :)
I'll happily change it if I have it wrong!

Reply:
TBH, I do not really know whether the character index is inclusive and I apologize for the noise. 

I just looked at the [D2D](https://github.com/wxWidgets/wxWidgets/blob/463deb03c90cf636cc2df7aa67d4d8afbddc5c64/src/msw/graphicsd2d.cpp#L3427) and [GDI+](https://github.com/wxWidgets/wxWidgets/blob/463deb03c90cf636cc2df7aa67d4d8afbddc5c64/src/msw/graphics.cpp#L2155) implementations where it seems the actual width of character 0 is stored at index 0. See also [the test](https://github.com/wxWidgets/wxWidgets/blob/cc931612eec2e3ea49200ebff45042135f3c3f9c/tests/graphics/measuring.cpp#L162), line 169.

Reply:
Ah, I didn't realize there was a test. (I searched for one but was not effective, I see.)

I'll make the change tomorrow. Thanks for bringing it up and the info! I'll get the test to pass.

Reply:
I have made the change. It turns out that test (sadly) doesn't test the graphics level code - it calls the DC level one, which uses a wxTextMeasure object instead. But that test at least showed me what the expected semantics are, and I'm grateful for that!

Reply:
Well, the test may not be difficult to add, e.g., simply something like [this](https://github.com/PBfordev/wxWidgets/commit/ff5758f4491e740cc58a6a97daa16a57a00ca03f) (but without a missing 's' in the method name which I just noticed).

However, I am not sure how useful it would be, particularly as it tests only the default renderer...